### PR TITLE
Add support for passing pathlib.Path objects as filenames

### LINKED ISSUE #1381: Support passing in pathlib.Path into pygmt modules
**Description of the desired feature**

_Originally posted by @liamtoney in https://github.com/GenericMappingTools/pygmt/issues/949#issuecomment-790796932_

> I'd add `Path` objects if they're not already supported. :)

The [`pathlib.Path`](https://docs.python.org/3.9/library/pathlib.html) is a cross-platform way of handling filepaths. Currently PyGMT supports passing in filenames as a `str`, but it would be good to ensure that `pathlib.Path` works too.

## What needs to be done :ballot_box_with_check:

1. This if-statement needs to be modified:

https://github.com/GenericMappingTools/pygmt/blob/11c13012c2c8a70006ac86d62e6f97f437451f82/pygmt/helpers/utils.py#L69-L70

2. Add some unit tests with `pathlib.Path` inputs, best placed in https://github.com/GenericMappingTools/pygmt/blob/master/pygmt/tests/test_info.py.

## How to make the change :monocle_face:

1. First, read up our contributing guidelines at https://www.pygmt.org/v0.4.0/contributing.html#editing-the-documentation
2. Leave a comment below, stating that you will tackle this issue. We will assign you to this issue and you can then start working on it.
3. To edit the utils.py file, go to https://github.com/GenericMappingTools/pygmt/edit/master/pygmt/helpers/utils.py and make the change!. Or you may want to do this in your code editor locally.
4. Submit a pull request, and be sure to fill it up with as much detail as possible.

**Are you willing to help implement and maintain this feature?** Happy to guide someone with some familiarity with Python Path objects to work on this. Will also be a great opportunity for someone wanting to learn how to write unit tests.

### PR DESCRIPTION
**Pull request by aitorres**
**Description of proposed changes**

This PR addresses and fixes issue #1381. By adding support to `pathlib.Path` values that represent file paths to the `data_kind` helper function, it can be used interchangeably with file name strings on `pygmt` modules.

Fixes #1381

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [x] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by weiji14** [comment]
Hmm, supporting pathlib.Path might not be as simple as I thought. We'll need to convert the `pathlib.Path` object into a `str` in order for the GMT C API to understand it. Do you want to walk through the `virtualfile_from_data` function at https://github.com/GenericMappingTools/pygmt/blob/master/pygmt/clib/session.py and see if you can find a good place to do the `pathlib.Path`->`str` conversion? Or we could change the line at https://github.com/GenericMappingTools/pygmt/blob/60a92b3ee2cc37cf9277a2fd623cffd2b52abcba/pygmt/helpers/utils.py#L106 to say `yield str(arg)` perhaps?

**Comment by aitorres** [comment]
I'm sorry for the delay! I've examined the `virtualfile_from_data` function and modified the conditional statement to cast `data` to `str` in presence of `"file"` data kind, instead of modifying the `dummy_context` function.

I've pushed my changes to my branch already, and will keep an eye on the tests in case something fails again. Thanks for your help!

File: pygmt/clib/session.py
Comment (weiji14) [inline pygmt/clib/session.py:8]: This `import pathlib` is unused and can be removed.

```suggestion
```

**Review by weiji14 — COMMENTED** [review]
> I'm sorry for the delay! I've examined the `virtualfile_from_data` function and modified the conditional statement to cast `data` to `str` in presence of `"file"` data kind, instead of modifying the `dummy_context` function.

Nice one, that's a much better solution! Will see if the tests pass on all platforms (Linux/macOS/Windows).

File: pygmt/clib/session.py
Comment (aitorres) [reply pygmt/clib/session.py]: Done!

**Review by weiji14 — APPROVED** [review]
Thanks @aitorres, looks good. I'll merge in 24hrs unless any of the other devs have something to add.

File: pygmt/helpers/utils.py
Comment (seisman) [inline pygmt/helpers/utils.py:74]: What will happen if people use `pathlib.PurePath` or `path.PurePosixPath`?

Comment (weiji14) [reply pygmt/helpers/utils.py:74]: The answer at https://stackoverflow.com/questions/58647584/how-to-test-if-object-is-a-pathlib-path/58966089#58966089 suggests using `isinstance(pathlib.PurPath)`. @aitorres, could you perhaps see if this works?

```suggestion
    if isinstance(data, (str, pathlib.PurePath)):
```

Also, it might be a good idea to test that `PureWindowsPath` and `PurePosixPath` objects work on Windows and macOS/Linux respectively. Let us know if you're able to update the unit test to do that. If not, we can also just ignore it.

Comment (weiji14) [reply pygmt/helpers/utils.py:74]: Gentle ping @aitorres to see if you're still available to finish up this PR! If not, we can get one of the maintainers to finish this up in the next week or so.

Comment (aitorres) [reply pygmt/helpers/utils.py]: Really sorry for the delay! I've just added and pushed the changes to use `PurePath`, and added two tests that specifically use `PureWindowsPath` and `PurePosixPath`.

File: pygmt/tests/test_info.py
Comment (seisman) [inline pygmt/tests/test_info.py:74]:
```suggestion
```

File: pygmt/tests/test_info.py
Comment (seisman) [inline pygmt/tests/test_info.py:33]: We can use `pytest.mark.parametrize` to pass different parameters to the same function so that we don't have to duplicate the code. The `pytest.mark.parametrize` is a little complicated here, because `pathlib.PureWindowsPath` only works on Windows and `pathlib.PurePosixPath` doesn't work on Windows. 
```suggestion
@pytest.mark.parametrize(
    "table",
    [
        pathlib.Path(POINTS_DATA),
        pytest.param(
            pathlib.PureWindowsPath(POINTS_DATA),
            marks=pytest.mark.skipif(
                sys.platform != "win32",
                reason="PureWindowsPath is only available on Windows",
            ),
        ),
        pytest.param(
            pathlib.PurePosixPath(POINTS_DATA),
            marks=pytest.mark.skipif(
                sys.platform == "win32",
                reason="PurePosixPath is not available on Windows",
            ),
        ),
    ],
)
def test_info_path(table):
```

File: pygmt/tests/test_info.py
Comment (seisman) [inline pygmt/tests/test_info.py:37]:
```suggestion
    output = info(table=table)
```

File: pygmt/tests/test_info.py
Comment (seisman) [inline pygmt/tests/test_info.py:5]:
```suggestion
import pathlib
import sys
```

**Comment by seisman** [comment]
Ping @aitorres to finalize this PR if you have time. Otherwise, I'll take over this PR and make edits to this branch.

**Review by weiji14 — APPROVED** [review]
Should be ok to merge after tests all pass, but prefer to have a second approval.

**Review by michaelgrund — APPROVED** [review]

**Pull request merged**
